Rubberduck.VBEEditor/ComManagement/ProjectsRepository.cs

```diff
@@ -0,0 +1,267 @@
+
+        private readonly ReaderWriterLockSlim _refreshProtectionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+
+        public ProjectsRepository(IVBE vbe)
+        {
+            _projectsCollection = vbe.VBProjects;
+        }
+
+        private void LoadCollections()
+        {
+            LoadProjects();
+            LoadComponentsCollections();
+            LoadComponents();
+        }
+
+        private void LoadProjects()
+        {
+            foreach (var project in _projectsCollection)
+            {
+                if (project.Protection == ProjectProtection.Locked)
+                {
+                    project.Dispose();
+                    continue;
+                }
+
+                EnsureValidProjectId(project);
+                _projects.Add(project.ProjectId, project);
+            }
+        }
+
+        private void EnsureValidProjectId(IVBProject project)
+        {
+            if (string.IsNullOrEmpty(project.ProjectId) || _projects.Keys.Contains(project.ProjectId))
+            {
+                project.AssignProjectId();
+            }
+        }
+
+        private void LoadComponentsCollections()
+        {
+            foreach (var (projectId, project) in _projects)
+            {
+                _componentsCollections.Add(projectId, project.VBComponents);
+            }
+        }
+
+        private void LoadComponents()
+        {
+            foreach (var components in _componentsCollections.Values)
+            {
+                LoadComponents(components);
+            }
+        }
+
+        private void LoadComponents(IVBComponents componentsCollection)
+        {
+            foreach (var component in componentsCollection)
+            {
+                var qmn = component.QualifiedModuleName;
+                _components.Add(qmn, component);
+            }
+        }
+
+        public void Refresh()
+        {
+            ExecuteWithinWriteLock(() => RefreshCollections());
+        }
+
+        private void ExecuteWithinWriteLock(Action action)
+        {
+            if (_disposed)
+            {
+                return; //The lock has already been diposed.
+            }
+
+            var writeLockTaken = false;
+            try
+            {
+                _refreshProtectionLock.EnterWriteLock();
+                writeLockTaken = true;
+                action.Invoke();
+            }
+            finally
+            {
+                if (writeLockTaken)
+                {
+                    _refreshProtectionLock.ExitWriteLock();
+                }
+            }
+        }
+
+        private void RefreshCollections()
+        {
+            //We save a copy of the collections and only refresh after the collections have been loaded again
+            //to avoid disconnecting any RCWs from the underlying COM object for objects that still exist.
```

Comment:
Do we need to wrap this body in a `try/catch`? I wonder because if we encounter an error after we've cleared some of the collections, we are going to end up dropping them to the floor. Should the `catch` restore the cleared collection to previous state to avoid problems w/ disconnected RCWs?

Reply:
Clearing the collections should actually never fail due to the write lock taken. The only things that can conceivably fail are loading the collections and disposing the wrappers. Maybe I could put a `try/finally` around loading the collections.